internal/plugin/external_plugins.go

```diff
@@ -0,0 +1,228 @@
+	}
+
+	plugins := Plugins{
+		HTTPPluginMap: map[string]http.Plugin{},
+		TCPPluginMap:  map[string]tcp.Plugin{},
+	}
+
+	for rawPluginName, rawPlugin := range rawPlugins {
+		logger.Infof("Loading plugin '%s'...", rawPluginName)
+
+		pluginType, pluginID, err := parsePluginMetadata(rawPlugin, rawPluginName, logger)
+		if err != nil {
+			logger.Errorln(err)
+			continue
+		}
+
+		switch pluginType {
+		case "connector.http":
+			httpPluginSym, err := getSymbolFromName(rawPlugin, "GetHTTPPlugin", logger)
+			if err != nil {
+				continue
+			}
+
+			httpPluginFunc, ok := httpPluginSym.(func() http.Plugin)
+			if !ok {
+				logger.Errorln(errors.New("GetHTTPPlugin could not be cast to the expected type"))
+				continue
+			}
+
+			plugins.HTTPPluginMap[pluginID] = httpPluginFunc()
+		case "connector.tcp":
+			tcpPluginSym, err := getSymbolFromName(rawPlugin, "GetTCPPlugin", logger)
+			if err != nil {
+				continue
+			}
+
+			tcpPluginFunc, ok := tcpPluginSym.(func() tcp.Plugin)
+			if !ok {
+				logger.Errorln(errors.New("GetTCPPlugin could not be cast to the expected type"))
+				continue
+			}
+
+			plugins.TCPPluginMap[pluginID] = tcpPluginFunc()
+		default:
+			logger.Errorln(fmt.Errorf("PluginInfo['type'] of '%s' is not supported", pluginType))
+			continue
+		}
+
+		logger.Warnf("Plugin %s/%s loaded", pluginType, pluginID)
+	}
+
+	return &plugins, nil
+}
+
+func getCheckedPlugins(pluginDir string,
+	checksumsFile string,
+	logger log.Logger,
+) ([]os.FileInfo, error) {
+
+	files, err := ioutil.ReadDir(pluginDir)
+	if err != nil {
+		return nil, err
+	}
+
+	if checksumsFile == "" {
+		logger.Warnln("Plugin hashes were not provided - tampering will not be detectable!")
+		return files, nil
+	}
+
+	// We override file listing if we did a verification to prevent additions
+	// to plugins between verification and loading the plugins.
+	if files, err = VerifyPluginChecksums(pluginDir, checksumsFile); err != nil {
+		logger.Errorln(err)
+		return nil, err
+	}
+
+	return files, nil
+}
+
+func loadPluginFiles(
+	pluginDir string,
+	files []os.FileInfo,
+	logger log.Logger,
+) (map[string]*go_plugin.Plugin, error) {
+
+	goPlugins := map[string]*go_plugin.Plugin{}
+	for _, file := range files {
+		fileName := file.Name()
+		if !strings.HasSuffix(fileName, ".so") {
+			logger.Warnf("File '%s' ignored as a plugin - missing appropriate extension", fileName)
+			continue
+		}
+
+		// Load shared library object
+		pluginObj, err := go_plugin.Open(fmt.Sprintf("%s/%s", pluginDir, file.Name()))
+		if err != nil {
+			logger.Errorln(err)
+			continue
+		}
+
+		logger.Infof("Adding '%s' as a plugin...", fileName)
+
+		goPlugins[fileName[:len(fileName)-3]] = pluginObj
+	}
+
+	return goPlugins, nil
+}
+
+func getSymbolFromName(rawPlugin *go_plugin.Plugin,
+	symbolName string,
+	logger log.Logger,
+) (go_plugin.Symbol, error) {
+
+	symbol, err := (*rawPlugin).Lookup(symbolName)
+	if err != nil {
+		logger.Errorln(err)
+	}
+	return symbol, err
+}
```

Comment:
Similar to Kumbi's comment, I def think these helper method should just return errors, and the logging responsibility should always be maintained in the highest level funcs.